src/algorithm/boundingbox.rs

```diff
@@ -0,0 +1,209 @@
+use num::{Float};
+
+use types::{Bbox, Point, MultiPoint, LineString, MultiLineString, Polygon, MultiPolygon};
+
+/// Calculation of the bounding box of a geometry.
+
+pub trait BoundingBox<T: Float> {
+    /// Return a Bounding Box of a geometry
+    ///
+    /// ```
+    /// use geo::{Point, LineString, Coordinate};
+    /// use geo::algorithm::boundingbox::BoundingBox;
+    ///
+    /// let mut vec = Vec::new();
+    /// vec.push(Point::new(40.02f64, 116.34));
+    /// vec.push(Point::new(42.02f64, 116.34));
+    /// vec.push(Point::new(42.02f64, 118.34));
+    /// let linestring = LineString(vec);
+    /// let bbox = linestring.bbox().unwrap();
+    ///
+    /// println!("Bbox top left coordinates: {}, {}", bbox.xmin, bbox.ymax);
+    /// println!("Bbox bottom right coordinates: {}, {}", bbox.xmax, bbox.ymin);
+    /// ```
+    ///
+    fn bbox(&self) -> Option<Bbox<T>>;
+}
+
+fn get_bbox<T>(vect: &Vec<Point<T>>) -> Option<Bbox<T>>
+    where T: Float
+{
+    if vect.is_empty() {
+        return None;
+    }
+    if vect.len() == 1 {
+        return Some(Bbox{xmin: vect[0].x(), ymax: vect[0].y(),
+                         xmax: vect[0].x(), ymin: vect[0].y()})
+    } else {
+        let (mut xmax, mut xmin) = (T::neg_infinity(), T::infinity());
+        let (mut ymax, mut ymin) = (T::neg_infinity(), T::infinity());
+        for pnt in vect.iter() {
+            let (px, py) = (pnt.x(), pnt.y());
+            if px > xmax {
+                xmax = px;
+            } else if px < xmin {
+                xmin = px;
+            }
+            if py > ymax {
+                ymax = py;
+            } else if py < ymin {
+                ymin = py;
+            }
```

Comment:
Dear Guillaume,

On Mon, Jul 18, 2016 at 9:23 AM, Guillaume P. <notifications@github.com>
wrote:

> In src/algorithm/boundingbox.rs
> <https://github.com/georust/rust-geo/pull/41#discussion_r71107118>:
>
> > +                         xmax: vect[0].x(), ymin: vect[0].y()})
> > +    } else {
> > +        let (mut xmax, mut xmin) = (T::neg_infinity(), T::infinity());
> > +        let (mut ymax, mut ymin) = (T::neg_infinity(), T::infinity());
> > +        for pnt in vect.iter() {
> > +            let (px, py) = (pnt.x(), pnt.y());
> > +            if px > xmax {
> > +                xmax = px;
> > +            } else if px < xmin {
> > +                xmin = px;
> > +            }
> > +            if py > ymax {
> > +                ymax = py;
> > +            } else if py < ymin {
> > +                ymin = py;
> > +            }
>
> You can abstract that as fn get_min_max(p: T, min: T, max: T) -> (T, T)
>

I'm trying to follow your comments, so at the moment the code looks like:

{{{

fn get_min_max<T>(p: T, min: T, max: T) -> (T, T)
    where T: Float
{
    if p > max {(min, p)} else if p < min {(p, max)} else {(min, max)}
}

fn get_bbox<T>(vect: &Vec<Point<T>>) -> Option<Bbox<T>>
    where T: Float
{
    if vect.is_empty() {
        return None;
    } else {
        let (mut xmax, mut xmin) = (vect[0].x(), vect[0].x());
        let (mut ymax, mut ymin) = (vect[0].y(), vect[0].y());
        for pnt in vect[1..].iter() {
            let (px, py) = (pnt.x(), pnt.y());
            (xmin, xmax) = get_min_max(px, xmin, xmax);
            (ymin, ymax) = get_min_max(py, ymin, ymax);
        }
        Some(Bbox{xmin: xmin, ymax: ymax,
                  xmax: xmax, ymin: ymin})
    }
}

}}}

but these line are problematic:

{{{
            (xmin, xmax) = get_min_max(px, xmin, xmax);
            (ymin, ymax) = get_min_max(py, ymin, ymax);
}}}

and the compailer complain with: error: invalid left-hand side expression
[E0070]

I'm in favor of a fold version to extract xmin, xmax, ymin ymax from a list
of points, but I want to avoid to cycle over the point more than one time.